File: 3/WaterArea.java
class WaterArea {
	
	public static int getWaterArea(int height[]) {
		int n = height.length;
		
		int leftMax[] = new int[n];
		for(int i=1; i<n; i++) {
			leftMax[i] = Math.max(height[i-1], leftMax[i-1]);
		}	

		int rightMax[] = new int[n];
		for(int i=n-2; i>=0; i--) {
			rightMax[i] = Math.max(height[i+1], rightMax[i+1]);
		}
		
		int waterArea = 0;
		for(int i=0; i<n; i++) {
			int minHeight = Math.min(leftMax[i], rightMax[i]);
			waterArea += height[i] < minHeight ? minHeight - height[i] : 0;	
		}
		
		return waterArea;
	}
	
	public static void main(String args[]) {
		int height[] = {0,8,0,0,5,0,0,10,0,0,1,1,0,3};
		
		System.out.println(getWaterArea(height));
	}
}

File: 3/js/src/WaterArea.ts
function waterArea(list: number[]): number {
	const n = list.length;

	const lMax: number[]  = [0];
	for(let i=1; i<n; i++) {
		lMax[i] = Math.max(lMax[i-1], list[i-1]);
	}

	const rMax: number[] = [];
	rMax[n-1] = 0;
	for(let i=n-2; i>=0; i--) {
		rMax[i] = Math.max(rMax[i+1], list[i+1]);
	}

	let area = 0;
	for(let i=0; i<n; i++) {
		const min = Math.min(lMax[i], rMax[i]);
		if(min > list[i]) {
			area += min - list[i];
		}
	}

	return area;
}

(function main() {
	const list = [0,8,0,0,5,0,0,10,0,0,1,1,0,3];
	console.log(waterArea(list));
})();
